File: day11/sol.py
import itertools as it
from functools import partial

from timing_util import Timing


def get_data(content: str) -> tuple[set[tuple[int, int]], list[int], list[int]]:
    grid = content.split("\n")

    galaxies = set()
    for x, line in enumerate(grid):
        for y, c in enumerate(line):
            if c == "#":
                galaxies.add((x, y))

    expansion_rows = []
    for x, line in enumerate(grid):
        if "#" not in line:
            expansion_rows.append(x)

    expansion_cols = []
    for y in range(len(grid[0])):
        if all(line[y] != "#" for line in grid):
            expansion_cols.append(y)

    return galaxies, expansion_rows, expansion_cols


def manhattan_distance(
    p1: tuple[int, int],
    p2: tuple[int, int],
    expansion_rows: list[int],
    expansion_cols: list[int],
    expansion_size: int,
) -> int:
    src_x, src_y = p1
    dst_x, dst_y = p2

    min_x = src_x if src_x < dst_x else dst_x
    max_x = src_x if src_x > dst_x else dst_x
    min_y = src_y if src_y < dst_y else dst_y
    max_y = src_y if src_y > dst_y else dst_y

    n_rows = 0
    for row in expansion_rows:
        if min_x < row < max_x:
            n_rows += 1

    n_cols = 0
    for col in expansion_cols:
        if min_y < col < max_y:
            n_cols += 1

    return (max_x - min_x) + (max_y - min_y) + (n_rows + n_cols) * expansion_size


def sol(data: tuple[set[tuple[int, int]], list[int], list[int]], expansion_size: int) -> int:
    galaxies, expansion_rows, expansion_cols = data
    return sum(
        manhattan_distance(g1, g2, expansion_rows, expansion_cols, expansion_size - 1)
        for g1, g2 in it.combinations(galaxies, 2)
    )


part1 = partial(sol, expansion_size=2)
part2 = partial(sol, expansion_size=1_000_000)


if __name__ == "__main__":
    with Timing("Elapsed time: "):
        with open("inputs/day11.txt") as in_file:
            data = get_data(in_file.read().strip())

        print(f"part1: {part1(data)}")
        print(f"part2: {part2(data)}")

File: day11/sol.go
package main

import (
	"fmt"
	"os"
	"strings"
	"time"
)

type Point struct {
	x, y int
}

func ReadFile(filename string) string {
	bytes, err := os.ReadFile(filename)
	if err != nil {
		panic(err)
	}
	return strings.TrimSpace(string(bytes))
}

func ParseInput(data string) ([]int, []int, []Point) {
	rows := make([]int, 0)
	cols := make([]int, 0)
	galaxies := make([]Point, 0)
	grid := strings.Split(data, "\n")
	for i, line := range grid {
		for j, val := range line {
			if val == '#' {
				galaxies = append(galaxies, Point{i, j})
			}
		}
	}
	for i := range grid {
		append_row := true
		for _, g := range galaxies {
			if i == g.x {
				append_row = false
				break
			}
		}
		if append_row {
			rows = append(rows, i)
		}
	}
	for j := range grid[0] {
		append_col := true
		for _, g := range galaxies {
			if j == g.y {
				append_col = false
				break
			}
		}
		if append_col {
			cols = append(cols, j)
		}
	}

	return rows, cols, galaxies
}

func Min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func Max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func Abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

func ManhattanDistance(a, b Point) int {
	return Abs(a.x-b.x) + Abs(a.y-b.y)
}

func solve(rows, cols []int, galaxies []Point, expansion int) int {
	galaxies_adjusted := make([]Point, len(galaxies))
	copy(galaxies_adjusted, galaxies)

	for i, g := range galaxies {
		n_r := 0
		for _, row := range rows {
			if g.x > row {
				n_r++
			}
		}
		galaxies_adjusted[i].x += (expansion - 1) * n_r

		n_c := 0
		for _, col := range cols {
			if g.y > col {
				n_c++
			}
		}
		galaxies_adjusted[i].y += (expansion - 1) * n_c
	}

	sum := 0
	for i := range galaxies_adjusted {
		for j := i + 1; j < len(galaxies_adjusted); j++ {
			sum += ManhattanDistance(galaxies_adjusted[i], galaxies_adjusted[j])
		}
	}
	return sum
}

func Part1(rows, cols []int, galaxies []Point) int {
	return solve(rows, cols, galaxies, 2)
}

func Part2(rows, cols []int, galaxies []Point) int {
	return solve(rows, cols, galaxies, 1_000_000)
}

func PrintTime(start time.Time, msg string) {
	dur := float64(time.Since(start).Nanoseconds())
	switch {
	case dur < 1e3:
		fmt.Printf("%v %.3fns\n", msg, dur)
	case dur < 1e6:
		fmt.Printf("%v %.3fµs\n", msg, dur/1e3)
	case dur < 1e9:
		fmt.Printf("%v %.3fms\n", msg, dur/1e6)
	default:
		fmt.Printf("%v %.3fs\n", msg, dur/1e9)
	}
}

func main() {
	start := time.Now()
	defer PrintTime(start, "Elapsed time:")

	text := ReadFile("inputs/day11.txt")
	rows, cols, data := ParseInput(text)

	fmt.Printf("Part1: %d\n", Part1(rows, cols, data))
	fmt.Printf("Part2: %d\n", Part2(rows, cols, data))
}
